### ISSUE TITLE
Add a better check for published dates when creating a book

### ISSUE THREAD

**Issue by knz13** [internetarchive/openlibrary#7900]
The problem is that when adding a new book to the library, the "When was it published?" doesn't have correct checking of valid dates especially in the days and months fields.

### Evidence / Screenshot (if possible)

<img width="1440" alt="Screen Shot 2023-05-23 at 20 18 07" src="https://github.com/internetarchive/openlibrary/assets/82895172/cb2e0eaa-068f-411f-b5df-3ecb6c1af93a">

<img width="1440" alt="Screen Shot 2023-05-23 at 20 18 56" src="https://github.com/internetarchive/openlibrary/assets/82895172/0b45b9d0-7ece-4ebc-a52b-057abb62d86b">

### Relevant url?
https://openlibrary.org/books/add

### Steps to Reproduce
1. Go to https://openlibrary.org/books/add
2. Add missing fields and check for dates such as -100/200/2002

* Actual: The field did not show any signs of an invalid output such as when trying with valid but future dates (such as 12/01/2026)
* Expected: An indication of a wrong date

### Details

- **Logged in (Y/N)?** Y
- **Browser type/version?** Safari Version 15.6.1 (17613.3.9.1.16)
- **Operating system?** MacOS
- **Environment (prod/dev/local)?** prod, dev and local

### Proposal & Constraints
Proposed solutions include changing the input to be a scroll wheel with dates or something more friendly for the user or using a regex such as "^(?:-?\d{1,4}|[1-9]\d{3})-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1\d|2[0-8])|(?:0[13-9]|1[0-2])-29|02-(?:0[1-9]|1\d|2[0-9])|02-30|02-31|(?:0[13578]|1[02])-(?:0[1-9]|1\d|2[0-9]|3[01])|(?:0[469]|11)-(?:0[1-9]|1\d|2[0-9]|30))$" to check for the input that could account for the appropriate dates (did not test the regex but I'm pretty sure it is right but does not account for leap years)

Here is an explanation written by chat gpt:
```
^                        # Start of the string
(?:                      # Non-capturing group
    -?                   # Optional minus sign for negative years
    \d{1,4}              # 1 to 4 digits for the year (allowing both 1 and 4 digit years)
    |                    # OR
    [1-9]\d{3}           # Match years from 1000 to 9999 (excluding 0000)
)
-                        # Hyphen separator between year and month
(?:                      # Non-capturing group for month and day
    (?:                  # Non-capturing group for months with 30 or 31 days
        0[1-9]|1[0-2]    # Months 01-12
        -                # Hyphen separator between month and day
        0[1-9]|1\d|2[0-8]   # Days 01-28
    )
    |                    # OR
    (?:                  # Non-capturing group for February
        0[13-9]|1[0-2]   # Months 03-12
        -                # Hyphen separator between month and day
        29               # Day 29 (allowing leap years)
    )
    |                    # OR
    02-                  # February
    (?:                  # Non-capturing group for days 01-28
        0[1-9]|1\d|2[0-9]   # Days 01-28
    )
    |                    # OR
    02-30                # Invalid day 30 for February
    |                    # OR
    02-31                # Invalid day 31 for February
    |                    # OR
    (?:                  # Non-capturing group for months with 31 days
        0[13578]|1[02]   # Months 01, 03, 05, 07, 08, 10, 12
        -                # Hyphen separator between month and day
        0[1-9]|1\d|2[0-9]|3[01]   # Days 01-31
    )
    |                    # OR
    (?:                  # Non-capturing group for months with 30 days
        0[469]|11        # Months 04, 06, 09, 11
        -                # Hyphen separator between month and day
        0[1-9]|1\d|2[0-9]|30   # Days 01-30
    )
)
$                        # End of the string

```

**Comment by LeadSongDog** [comment]
Further simple checks are possible:
-Exclude Jan 1 dates, they are almost all fake data, just keep the year
-Ensure pub date is after last of authors’ birth dates (if known)

**Comment by mheiman** [comment]
This seems like an easy win, but in the book world, dates are notoriously tricky. Simple year-month-day validation won't actually do the job.

We get most of our bibliographic data from MARC records — take a look at the specifications for how dates are represented there:  https://www.loc.gov/marc/bibliographic/bd008a.html

**Comment by LeadSongDog** [comment]
@mheiman The date formats for MARC21 were updated in 2021. Various format codings are supported, notably including the Feb 2019 version of EDTF: https://www.loc.gov/standards/datetime/

We decided some time back that EDTF should be the internal format on OL. There are lots of extant git repos with parsers to choose from, there’s no reason to roll our own.

**Comment by scottbarnes** [comment]
I am proposing the following to try to tease out people's views so we can get some sort of consensus about how, or if, we should address this.

As patrons use https://openlibrary.org/books/add to manually add books, what about a two-part solution whereby we first update the input text for the `publish_date` field to give a bit more guidance as to the format, e.g.: 
`<input type="text" class="required publish-date" id="publish_date" name="publish_date" placeholder="Format as year-month-day, if possible">`
![image](https://github.com/internetarchive/openlibrary/assets/26524678/fc2b2591-a76f-45eb-92cb-4ca00eec80ce)

Then we update the JavaScript to:
1. check for a facially invalid date using some existing JavaScript library;
2. when the patron clicks add, show an alert that the date looks suspect; and
3. allow the patron to override the alert and enter the date they wish to enter, regardless of validity.

In this way the `publish_date` field might be like the `id_value` field:
![image](https://github.com/internetarchive/openlibrary/assets/26524678/8e0c76d1-c001-4f03-9ee7-b06a9fee5022)

**Comment by tfmorris** [comment]
More guidance for users is good, but many publication dates are just years, so I'd suggest something like "Format as year or numeric year-month-day." 

It would also be useful to align this with the input standards for other dates. I believe that the author birth and death dates currently suggests a different format. That's just going to confuse users.

**Comment by seabelis** [comment]
It has already come up in past conversations that a date picker is not especially useful for this field. There's nothing user-unfriendly about just typing or pasting a date. The date field needs to accommodate these formats, https://www.loc.gov/standards/datetime/, as @LeadSongDog has pointed out. Additional guidance is fine, but don't make me (and everyone else) use a date picker or some other forced formats that don't always apply.

**Comment by knz13** [comment]
I agree that it is not user-unfriendly the way it is being handled now, but as someone that never really read many books it would be helpful to have some directions about how to format the date on the actual page for adding the book. Maybe it could be more clear to have a tooltip when hovering over the "When was it published?" title or a question mark button with a link to the page with the format guidance?

Maybe some text like: "Here's some more information on how to format the dates: https://www.loc.gov/standards/datetime/"

**Comment by LeadSongDog** [comment]
How about the tip "YYYY, YYYY-MM, or for special cases, see https://www.loc.gov/standards/datetime/"

**Comment by seabelis** [comment]
We can link to the help section here, https://openlibrary.org/help/faq/editing#publication-date.
